### PR TITLE
feat: add sflog (Salesforce debug log) parser and highlights

### PR DESCRIPTION
**Pull request by xixiaofinland**

### CODE REVIEW

File: queries/sflog/highlights.scm
Comment (lucario387) [inline queries/sflog/highlights.scm:1]: Queries need to be adapted

File: queries/sflog/highlights.scm
Comment (lucario387) [inline queries/sflog/highlights.scm:6]: version is a string, so maybe `@string.special` will fit?

Comment (xixiaofinland) [reply queries/sflog/highlights.scm:6]: Thanks for the help! I'm still learning what @ fits what node best.
Updating it.

Comment (xixiaofinland) [reply queries/sflog/highlights.scm]: @lucario387 it seems actions all passed. Pls let me know if any more actions are required on my end :)

File: lua/nvim-treesitter/parsers.lua
Comment (clason) [inline lua/nvim-treesitter/parsers.lua:1917]: Did @aheber volunteer to maintain _the queries_ in _this repo_?

Comment (xixiaofinland) [reply lua/nvim-treesitter/parsers.lua:1917]: @clason  yes, I'm in closely contact with him, and agreed upon this with him.

Comment (aheber) [reply lua/nvim-treesitter/parsers.lua:1917]: Yes, this can have my name on it.

Comment (clason) [reply lua/nvim-treesitter/parsers.lua:1917]: Well, it's less about whose name is on it but who I can ping about issues and PRs (and breaking changes) and expect a response...

Comment (xixiaofinland) [reply lua/nvim-treesitter/parsers.lua:1917]: This lang parser sits in the repo owned by @aheber, but I'm the one who would actively respond to pings about it.
Shall I update the name to my name @clason ?

Comment (clason) [reply lua/nvim-treesitter/parsers.lua:1917]: Yes (and same for the other parsers in that repo, if @aheber is also not maintaining those queries here).

Comment (clason) [reply lua/nvim-treesitter/parsers.lua]: As long as he's available for questions about them, that's OK then. (We've had very bad experiences with maintainers who basically stopped responding as soon as the PR was merged; I want to avoid that going forward.)

Comment (clason) [reply lua/nvim-treesitter/parsers.lua]: So, yes, if _you_ are responsible for the queries, then _you_ should be the maintainer. (Since the parser is not developed in this repo, this is not very relevant.)

Comment (xixiaofinland) [reply lua/nvim-treesitter/parsers.lua]: This won't be the case on our parsers, even if we don't work in the Salesforce ecosystem in the future, we will make sure somebody taking them over. 

And I have just updated apex, sosl, soql, sflog maintainers to my name.

Comment (clason) [reply lua/nvim-treesitter/parsers.lua]: As I said, I'm more concerned about the _queries_ ;) (But that is good to hear; we've also had problems with abandoned parsers which required adaptation to upstream tree-sitter changes to keep working...)

Comment (xixiaofinland) [reply lua/nvim-treesitter/parsers.lua]: I'm curious, do you get users requesting query updates in this repo? I think they should ask the listed maintainers.

I'm asking because I'd be so happy to see requests in our parsers. It means there are real people using and caring about them 😸

Comment (clason) [reply lua/nvim-treesitter/parsers.lua]: Yes, since the queries are editor-specific and curated to be consistent (as far as possible) across languages. For issues that require parser changes, we push to report them at the parser repo and close them here.

Comment (xixiaofinland) [reply lua/nvim-treesitter/parsers.lua]: Got it, all clear. Updating the query knowledge is something I never find the excuse to take action, hopefully one day driven by the user requests ;).

Comment (aheber) [reply lua/nvim-treesitter/parsers.lua]: I'm happy for @xixiaofinland to be the listed maintainer here. I'll gladly help with any query stuff or parser stuff that comes up and I've been responsive here in the past when my parser change has broken these queries, preventing updates.
I do plan to keep that going but have seen enough languishing projects to understand the concern.

Is there a way to provide a secondary or backup maintainer?

Comment (clason) [reply lua/nvim-treesitter/parsers.lua]: (Nothing of that was personal, mind you; just explaining why I am pressing on that point.)

And the maintainers field is a table for a reason ;)

Comment (aheber) [reply lua/nvim-treesitter/parsers.lua]: Thanks, @clason!

@xixiaofinland, would you be interested in having us both as listed maintainers? 

Changing the relevant sections to: 
`maintainers = { "@xixiaofinland", "@aheber" }`

That would help get any problems resolved more quickly and add more long-term stability.

Comment (xixiaofinland) [reply lua/nvim-treesitter/parsers.lua]: @aheber let's do this, I added both of our names :)

File: queries/sflog/highlights.scm
Comment (ribru17) [inline queries/sflog/highlights.scm:27]: I think this should be removed, since more specific captures below apply to it

File: queries/sflog/highlights.scm
Comment (ribru17) [inline queries/sflog/highlights.scm:1]: Can we get `","`, `";"`, and `":"` as `@punctuation.delimiter`, and `"|["`, `"]|"`, `"|("`, `")|"`, `"("` and `")"` as `@punctuation.bracket`?

Maybe `"|"` as `@operator` (or whatever you feel suits it best)?

File: queries/sflog/highlights.scm
Comment (ribru17) [inline queries/sflog/highlights.scm:37]: I think the outer `@number` highlight should be removed, child queries already highlight over all function areas

File: queries/sflog/highlights.scm
Comment (ribru17) [inline queries/sflog/highlights.scm:22]: All of these number highlights are redundant, since they are covered above

File: queries/sflog/highlights.scm
Comment (ribru17) [inline queries/sflog/highlights.scm:44]: [`event_identifier` just maps immediately to `identifier`](https://github.com/aheber/tree-sitter-sfapex/blob/caffc6ebd1ab16ba4d4dec367cebb03ca2411872/sflog/grammar.js#L81), so we should remove either `@keyword` or `@type` (probably `@keyword` imo)

File: queries/sflog/highlights.scm
Comment (ribru17) [inline queries/sflog/highlights.scm:1]: Should `anonymous_block` be given a highlight? string maybe?

File: queries/sflog/highlights.scm
Comment (ribru17) [inline queries/sflog/highlights.scm:19]: I think this whole query should be removed: number is already covered and other than that we just have a few parenthesis which should eventually be covered as `@punctuation.bracket`

File: queries/sflog/highlights.scm
Comment (ribru17) [inline queries/sflog/highlights.scm:40]: This should be removed, since location will be either a `(number)` (already covered) or `"EXTERNAL"` (which should be its own thing)

File: queries/sflog/highlights.scm
Comment (ribru17) [inline queries/sflog/highlights.scm:8]: This is redundant since `(namespace)` just maps directly to `(identifier)`

File: queries/sflog/highlights.scm
Comment (ribru17) [inline queries/sflog/highlights.scm:1]: Maybe `"out of"` as `@keyword`? Your call, I think it's also fine as `@property`

File: queries/sflog/highlights.scm
Comment (xixiaofinland) [reply queries/sflog/highlights.scm:44]: good point!

File: queries/sflog/highlights.scm
Comment (xixiaofinland) [reply queries/sflog/highlights.scm:1]: I added the @punctuation.delimiter and @punctuation.bracket, but not sure if there are redundant ones, for example as it has "|" and "(", does it still need "|("? 
Also "]|" would throw error, why?

Comment (ribru17) [reply queries/sflog/highlights.scm:1]: In this case they are not redundant, because parser captured them separately as "|(", not "|" then "(". If you do :InspectTree then hit "a" in the tree window you can see all the anonymous nodes. Also I can check out the error soon, afk right now. Thanks for making the other changes!

Comment (xixiaofinland) [reply queries/sflog/highlights.scm:1]: Not at all, I should definitely thank you Riley. I'm bad at defining queries, it's really good to learn and rectify those points from you.

**Comment by clason** [comment]
@aheber I took a look at the grammar; these rules
https://github.com/aheber/tree-sitter-sfapex/blob/caffc6ebd1ab16ba4d4dec367cebb03ca2411872/sflog/grammar.js#L11
https://github.com/aheber/tree-sitter-sfapex/blob/caffc6ebd1ab16ba4d4dec367cebb03ca2411872/sflog/grammar.js#L41
are unsafe (current tree-sitter CLI master has changed behaviour around such single-child rules). You should remove the `seq` to be safe and simplify the parser.

**Comment by aheber** [comment]
@clason, good catch. Fixed.

File: queries/sflog/highlights.scm
Comment (ribru17) [reply queries/sflog/highlights.scm:1]: Of course :) The queries were not bad at all. I think all issues are resolved now, I was wrong about `"]|"`. :+1:

**Comment by ribru17** [comment]
@xixiaofinland would you mind updating the `lockfile.json` to match the parser update? Then I think this is good to go. Thank you both!

**Review by ribru17 — APPROVED** [review]

**Comment by ribru17** [comment]
Also @aheber, this is not blocking but would it be possible to expose the text [here](https://github.com/aheber/tree-sitter-sfapex/blob/c47a639e1a0e5407ee39a8761ec1accebe6dacfc/sflog/grammar.js#L41) as maybe `(execute_anonymous)`, `":"`, `(apex_code)` (or something equivalent)? It would be nice because then we could inject apex into these locations

**Comment by aheber** [comment]
I'll look at it.

I didn't do it yet because I need to see how it behaves in a few other contexts. Particularly language.

I'd love to find a way to separate the leading text and treat all the apex lines together.

I'll improve it this weekend.

**Comment by ribru17** [comment]
Thanks, no worries take your time

**Comment by aheber** [comment]
Done. You can see a more elaborate example here:

https://github.com/aheber/tree-sitter-sfapex/blob/9fe5d1fbfd75c11ba48a32db27ac0ed5da57ec78/sflog/test/corpus/anonymous.txt#L41

I don't know how well the embeddings will work on that but that is an extreme example, most such blocks will be much more reasonable about complete statements on a single line.

**Comment by xixiaofinland** [comment]
@ribru17  do I still need to update `lockfile.json` as it's already merged?

**Comment by clason** [comment]
No, that happens automatically with a daily cronjob.

**Comment by xixiaofinland** [comment]
@clason  Does it mean, in this case the new version aheber created 2hr ago would be available in nvim-treesitter within 24hrs done by the daily cronjob?

I'm eager to try out the embedded `apex` in `sflog` feature :)

**Comment by clason** [comment]
Yes. (The job runs around ~7:00 UTC and is merged automatically if it passes CI.)

You can also run the `./scripts/update-lockfile.sh` locally and then do `:TSUpdate` if you're impatient.

**Comment by xixiaofinland** [comment]
> Done. You can see a more elaborate example here:
> 
> https://github.com/aheber/tree-sitter-sfapex/blob/9fe5d1fbfd75c11ba48a32db27ac0ed5da57ec78/sflog/test/corpus/anonymous.txt#L41
> 
> I don't know how well the embeddings will work on that but that is an extreme example, most such blocks will be much more reasonable about complete statements on a single line.

You should start using Neovim ;).
I'm not sure consistent `apex` in log is, we definitely needs more granuar-level in order to separate the code from the entire log line (in the future)

Image 010

**Pull request merged**
